### PR TITLE
ODH-ADR-0011: Add RayTune for providing HPO functionality

### PR DESCRIPTION
**Pull request by dinogun**
Proposal to add RayTune to provide Hyper Parameter Optimization support

## Description
This document proposes the integration of [RayTune](https://docs.ray.io/en/latest/tune/index.html) Hyper Parameter Optimization (HPO) Library as the primary choice for HPO for OpenShift AI.

### CODE REVIEW

File: ODH-ADR-0011-hpo-raytune.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+| odh.HPOConfig         | Artifact      | Captures the hyperparameter configuration for all trails of HPO experiments. | New  |
+| odh.HPOTrial          | Context       | Provides information about individual trials conducted within an experiment, aiding in experiment management. | New   |
+| odh.HPOExperiment     | Context       | Serves as a parent context for trials, facilitating the organiszation and comparison of experiments.          | New   |
+
+
+## Alternatives
+
+RayTune and Katib were evaluated for this proposal
+1. RayTune supports a few additional algorithms that are not part of the Katib at this time. This includes Ax (Adaptive Experimentation), Dragonfly, BlendSearch, CFO (Cost Frugal Hyperparamater Optimization), HEBO (Heteroscedastic Evolutionary Bayesian Optimization), Nevergrad, ZOOpt, SigOpt and Repeater.
+1. RayTune integrates seamlessly into OpenShift AI workflow as it is part of the Ray framework that is currently in use by OpenShift AI. It does not require any additional setup. HPO trials run as processes on Ray clusters.
```
Comment (terrytangyuan) [inline ODH-ADR-0011-hpo-raytune.md:70]: Would RayTune integrate well if we are using Kubeflow training-operator?

File: ODH-ADR-0011-hpo-raytune.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+| odh.Metrics           | Artifact      | Stores evaluation metrics generated during model training or inference.  | New  |
+| odh.HPOConfig         | Artifact      | Captures the hyperparameter configuration for all trails of HPO experiments. | New  |
+| odh.HPOTrial          | Context       | Provides information about individual trials conducted within an experiment, aiding in experiment management. | New   |
+| odh.HPOExperiment     | Context       | Serves as a parent context for trials, facilitating the organiszation and comparison of experiments.          | New   |
+
+
+## Alternatives
+
+RayTune and Katib were evaluated for this proposal
+1. RayTune supports a few additional algorithms that are not part of the Katib at this time. This includes Ax (Adaptive Experimentation), Dragonfly, BlendSearch, CFO (Cost Frugal Hyperparamater Optimization), HEBO (Heteroscedastic Evolutionary Bayesian Optimization), Nevergrad, ZOOpt, SigOpt and Repeater.
```
Comment (terrytangyuan) [inline ODH-ADR-0011-hpo-raytune.md:69]: Are these supported by the frameworks supported by Katib? e.g. hyperopt, optuna, etc. https://github.com/kubeflow/katib?tab=readme-ov-file#search-algorithms

File: ODH-ADR-0011-hpo-raytune.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+Using RayTune does not need any additional work to support multi-tenancy. Since RayTune trials run as jobs in a worker pod within the same namespace, the OpenShift RBAC currently in effect on a given OpenShift AI Project is sufficient to isolate the pods from other OpenShift AI namespaces / projects.
+
+
+## Stakeholder Impacts
+
+| Group                         | Key Contacts     | Date       | Impacted? |
+| ----------------------------- | ---------------- | ---------- | --------- |
+
+
+## References
```
Comment (terrytangyuan) [inline ODH-ADR-0011-hpo-raytune.md:95]: Have you seen this paper that outlines some advantages of Katib over the alternatives? https://arxiv.org/abs/2006.02085

image

File: ODH-ADR-0011-hpo-raytune.md (hunk 2 above)
Comment (dinogun) [reply ODH-ADR-0011-hpo-raytune.md:70]: Both RayTune and Katib support all of the basic frameworks such as hyperopt, optuna.
Katib supports two algorithms that are currently not supported by RayTune, viz, Katib ENAS and DARTS.
RayTune OTOH supports several algorithms that are not currently available in Katib: Ax, Dragonfly, Blendsearch, Cost Frugal HPO, HEBO, Nevergrad, ZOOpt, SigOpt ...

File: ODH-ADR-0011-hpo-raytune.md (hunk 1 above)
Comment (dinogun) [reply ODH-ADR-0011-hpo-raytune.md:71]: RayTune works independent of the Kubeflow training operator, have not explored the impact of training operator on Raytune

File: ODH-ADR-0011-hpo-raytune.md (hunk 3 above)
Comment (dinogun) [reply ODH-ADR-0011-hpo-raytune.md:96]: Please note that the intent is to support both RayTune and Katib eventually into OpenShift AI. However it appears that more work is needed at this time to be able to integrate Katib with OAI at this time. Eg, see this [script](https://github.com/kruize/hpo-poc/blob/main/koroai/koroai), which is needed to run Katib in OpenShift AI

**Comment by etirelli** [comment]
@dinogun is this ADR ready for merge? who are the stakeholders that need to approve it?

**Pull request closed without merging**